share/h2o/mruby/dos_detector.rb

```diff
@@ -0,0 +1,109 @@
+require "lru_cache.rb"
+
+class DoSDetector
+
+  def initialize(config={})
+    config = {
+      :strategy   => CountingStrategy.new,
+      :callback   => self.class.default_callback,
+      :forwarded  => true,
+      :cache_size => 128,
+    }.merge(config)
+
+    @strategy = config[:strategy]
+    @callback = config[:callback]
+    @forwarded = !!(config[:forwarded])
+    @cache = LRUCache.new(config[:cache_size])
+    raise "strategy must not be nil" if @strategy.nil?
+    raise "callback must not be nil" if @callback.nil?
+  end
+
+  def self.default_callback
+    Proc.new do |vars|
+      [ 403, { "Content-Type" => "text/plain" }, [ "Forbidden" ] ]
+    end
+  end
+
+  def self.fallthrough_callback
+    Proc.new do |vars|
+      env_headers = vars.map { |k, v| [ "x-fallthru-set-dos-#{k}", v.to_s ] }.to_h
+      [ 399, env_headers, [] ]
+    end
+  end
+
+  def call(env)
+    now = Time.now.to_i
+
+    ip = env['REMOTE_ADDR']
+    if @forwarded && (xff = env['HTTP_X_FORWARDED_FOR'])
+      ip = xff.split(",")[0]
+    end
+
+    unless client = @cache.get(ip)
+      client = { :ip => ip }
+      @cache.set(ip, client)
+    end
+
+    detected, vars = @strategy.detect?(client, now, env)
+    if detected
+      vars = { :ip => client[:ip] }.merge(vars || {})
+      return @callback.call(vars)
+    end
+    return [ 399, {}, [] ]
```

Comment:
How about invoking the callback regardless of if an attack was detected?

I believe it would be better to do in each callback check the `detected` flag and perform appropriate action (return 399, 403, etc.), since doing so would reduce the code paths that needs to be taken care of when writing a custom callback.